screensaver.turnoff/screensaver.py

```diff
@@ -1,164 +1,300 @@
+         function='run_command',
+         args_off=['su', '-c', 'echo 0 >/sys/devices/virtual/graphics/fb0/cec'],
+         args_on=['su', '-c', 'echo 1 >/sys/devices/virtual/graphics/fb0/cec']),
+    # NOTE: Contrary to what one might think, 1 means off and 0 means on
+    dict(name='backlight-rpi', title='Backlight on Raspberry Pi (kernel)',
+         function='run_command',
+         args_off=['su', '-c', 'echo 1 >/sys/class/backlight/rpi_backlight/bl_power'],
+         args_on=['su', '-c', 'echo 0 >/sys/class/backlight/rpi_backlight/bl_power']),
+    dict(name='backlight-odroid-c2', title='Backlight on Odroid C2 (kernel)',
+         function='run_command',
+         args_off=['su', '-c', 'echo 0 >/sys/class/amhdmitx/amhdmitx0/phy'],
+         args_on=['su', '-c', 'echo 1 >/sys/class/amhdmitx/amhdmitx0/phy']),
+]
+
+POWER_METHODS = [
+    dict(name='do-nothing', title='Do nothing',
+         function='log', kwargs_off=dict(level=2, msg='Do nothing to power off system')),
+    dict(name='suspend-builtin', title='Suspend (built-in)',
+         function='jsonrpc', kwargs_off=dict(method='System.Suspend')),
+    dict(name='hibernate-builtin', title='Hibernate (built-in)',
+         function='jsonrpc', kwargs_off=dict(method='System.Hibernate')),
+    dict(name='quit-builtin', title='Quit (built-in)',
+         function='jsonrpc', kwargs_off=dict(method='Application.Quit')),
+    dict(name='shutdown-builtin', title='ShutDown action (built-in)',
+         function='jsonrpc', kwargs_off=dict(method='System.Shutdown')),
+    dict(name='reboot-builtin', title='Reboot (built-in)',
+         function='jsonrpc', kwargs_off=dict(method='System.Reboot')),
+    dict(name='powerdown-builtin', title='Powerdown (built-in)',
+         function='jsonrpc', kwargs_off=dict(method='System.Powerdown')),
+]
+
+
+class SafeDict(dict):
+    ''' A safe dictionary implementation that does not break down on missing keys '''
+    def __missing__(self, key):
+        ''' Replace missing keys with the original placeholder '''
+        return '{' + key + '}'
+
+
+def from_unicode(text, encoding='utf-8'):
+    ''' Force unicode to text '''
+    if sys.version_info.major == 2 and isinstance(text, unicode):  # noqa: F821; pylint: disable=undefined-variable
+        return text.encode(encoding)
+    return text
+
+
+def to_unicode(text, encoding='utf-8'):
+    ''' Force text to unicode '''
+    return text.decode(encoding) if isinstance(text, bytes) else text
+
 
+def log(level=1, msg='', **kwargs):
+    ''' Log info messages to Kodi '''
+    max_log_level = int(get_setting('max_log_level', 0))
+    if not DEBUG_LOGGING and not (level <= max_log_level and max_log_level != 0):
+        return
+    from string import Formatter
+    if kwargs:
+        msg = Formatter().vformat(msg, (), SafeDict(**kwargs))
+    msg = '[{addon}] {msg}'.format(addon=ADDON_ID, msg=msg)
+    xlog(from_unicode(msg), level % 3 if DEBUG_LOGGING else 2)
```

Comment:
Your logging is still weird. Add-ons should only log use the xbmc.LOGDEBUG (value 0) level. Seems that logging happens at all levels except 3 (because of the %3).

Reply:
Only when you enable debug logging.

Reply:
If I don't have `DEBUG_LOGGING` on False, you always log with 2 = LOGNOTICE not LOGDEBUG (0)

Reply:
So if you have `DEBUG_LOGGING` set to False, it depends on `max_log_level` whether anything gets logged. Specifically for **screensaver.turnoff** by default we log information regarding which display-method and/or system-method was attempted for suspend/resume or shutdown.

This is not different than how it used to work before. We need to log what action was performed in both cases even when `DEBUG_LOGGING` is disabled, because we offer about 10 different possible display methods and a handful of shutdown methods. We cannot know which method will work for the end-user as it not only depends on the system or platform, but also depends on the display device (projector, TV, display, smartphone, touchscreen, ...)

The user can disable logging, or enable more verbosity in the screensaver settings to help him troubleshoot his scenario. And the reason that this is required without `DEBUG_LOGGING` is because we need to know exactly what method was used when it failed, which could be at any time, even when Kodi is not being used. We cannot expect our users to have Kodi debug logging enabled all the time, or enable it before the screensaver kicks in.